### ISSUE TITLE
AUROC runs into memory issues

### ISSUE THREAD

**Issue by andrew-bydlon** [Lightning-AI/metrics#1461]
Just wanted to share my finding on this particular metric.  Basically:
1. AUROC stores all states that are entered into it:
```
In [18]: vars(met.MetricsDict['AUC'])['target'][0]
Out[18]: 
tensor([[-1,  1,  1,  1, -1,  1,  1,  1,  0],
        [-1,  1, -1,  1, -1,  0, -1, -1, -1],
        ...
        [-1, -1, -1,  1,  0,  0,  0, -1,  1],
        [ 1,  0,  0, -1,  0,  0,  0,  0, -1]], device='cuda:0',
       dtype=torch.int32)

In [19]: vars(met.MetricsDict['AUC'])['preds'][0]
Out[19]: 
tensor([[0.4803, 0.4768, 0.5145, 0.4828, 0.4813, 0.5133, 0.5086, 0.5011, 0.5435],
        [0.4808, 0.4781, 0.5145, 0.4825, 0.4808, 0.5129, 0.5086, 0.5013, 0.5436],
        ...
        [0.4814, 0.4769, 0.5151, 0.4829, 0.4813, 0.5122, 0.5085, 0.5019, 0.5427],
        [0.4810, 0.4787, 0.5141, 0.4828, 0.4813, 0.5128, 0.5096, 0.5027, 0.5445]],
       device='cuda:0')
```

2. If you are using these metrics in a training loop (as I am), then it will store the actual gradients of the predictions unless you explicitly detach:

```
In [21]: for i in range(1000):
    ...:     predictions=model(torch.randn((32,991,20), device='cuda')); labels = torch.randint(-1,2,(32,9), device='cuda')
    ...:     met(predictions.detach(),labels)
```
(Results in 4MB increase in GPU memory)

```
In [22]: for i in range(10):
    ...:     predictions=model(torch.randn((32,991,20), device='cuda')); labels = torch.randint(-1,2,(32,9), device='cuda')
    ...:     met(predictions,labels)
    ...: 
```
(Results in a 120 MB increase in GPU memory with a small model).

I got around this issue by detaching mid loop for the metrics class:
```
if self.mixed_precision:
	with autocast(device_type='cuda', dtype=torch.float16):
		predictions = self.model(feats)
		loss = self.loss_function(predictions=predictions, labels=labels)
		self.train_metric(predictions.detach(), labels)
	self.scaler.scale(loss).backward()
	self.scaler.step(self.optimizer)
	self.scaler.update()
```
but would recommend:
1. Detaching for AUROC (and any metric that stores targets/predictions) by default.
2. Coming up with an averaging method that doesn't require state storage, perhaps just a running mean, for compute.

**Comment by SkafteNicki** [comment]
Hi @andrew-bydlon,

Thanks for opening this issue. Some comments:

1. From v0.10 of torchmetrics metrics such as `AUROC` support both binned mode and non-binned mode. The default is non-binned mode which is the most accurate way to calculate the metric but is not memory efficient because all predictions and targets needs to be saved. However, if binned mode is used then the memory is constant. This can be enabled by setting the `thresholds` argument
```python
import torchmetrics, torch
metric = torchmetrics.classification.BinaryAUROC(thresholds=100)
metric(torch.rand(100,), torch.randint(2, (100,)))
metric.confmat.shape # prints (100, 2, 2)
metric(torch.rand(100,), torch.randint(2, (100,)))
metric.confmat.shape # prints (100, 2, 2)
```
shape of the saved state is constant in this case.

2. For you second point, we do a detach if you use the `.update` method. However, if you call `forward` as in your example we do not detach because some users would like to use the metrics as loss functions and thus needs the gradients. Thus for you example where it seems you are just interested in accumulating the metric over multiple batches, please use the `.update` method instead of `forward`

**Comment by andrew-bydlon** [comment]
Thank you for these tips.  This does seem to solve my problem.  I'll use update and fixed threshold bins going forward.
